app/controllers/commitments_controller.rb

```diff
@@ -31,4 +31,74 @@ def list
     @commitments = Commitment.paginate_commitments(params.to_json)
     render json: @commitments
   end
+
+  def new
+    @commitment = Commitment.new(criterium_id: params[:criterium_id])
+    # set @form_hash for use by vue component here
+  end
+
+  def create
+    @commitment = Commitment.new(commitment_params)
+    if @commitment.save
+      respond_to do |format|
+        format.json { json_response({ commitment: @commitment }, :created) }
+      end
+    else
+      respond_to do |format|
+        format.json {
+          error_messages = @commitment.errors.messages.dup
+          @commitment.state = :draft
+          @commitment.save
+          json_response({ errors: error_messages }, :unprocessable_entity) 
+        }
+      end
+    end
+  end
+
+  def edit
+    # set @form_hash for use by vue component here
+  end
+
+  def update
+    if @commitment.update(commitment_params)
+      respond_to do |format|
+        format.json { json_response({ commitment: @commitment }, 204) }
+      end
+    else
+      respond_to do |format|
+        format.json { json_response({ errors: @commitment.errors }, :unprocessable_entity) }
+      end
+    end
+  end
+
+  private
+
+  def set_commitment
+    @commitment = Commitment.find(params[:id])
+  end
+
+  def commitment_params
+    params.require(:commitment).permit(
+      :latitude,
+      :longitude,
+      :current_area_ha,
+      :committed_year,
+      :implementation_year,
+      :name,
+      :description,
+      :duration_years,
+      :stage,
+      :responsible_group,
+      :state,
+      :joint_governance_description,
+      :geospatial_file,
+      country_ids: [],
+      action_ids: [],
+      threat_ids: [],
+      manager_ids: [],
+      objective_ids: [],
+      links_attributes: [:url, :name],
+      progress_documents_attributes: [:id, :progress_notes, :document]
```

Comment:
Small thing, but is there a reason for this order of params? Should they be alphabetical or form-chronological?

Reply:
i've reordered them. closer to alphabetical, but keeping associations at the end, and for association paams keeping id first, and destroy last